engine/consensus/approvals/core.go

```diff
@@ -0,0 +1,336 @@
+	verifier                             module.Verifier                 // used to validate result approvals
+	seals                                mempool.IncorporatedResultSeals // holds candidate seals for incorporated results that have acquired sufficient approvals; candidate seals are constructed  without consideration of the sealability of parent results
+	approvalConduit                      network.Conduit                 // used to request missing approvals from verification nodes
+	requestTracker                       *sealing.RequestTracker         // used to keep track of number of approval requests, and blackout periods, by chunk
+}
+
+func NewApprovalProcessingCore(headers storage.Headers, state protocol.State, assigner module.ChunkAssigner,
+	verifier module.Verifier, seals mempool.IncorporatedResultSeals, approvalConduit network.Conduit, requiredApprovalsForSealConstruction uint, emergencySealingActive bool) (*approvalProcessingCore, error) {
+
+	lastSealed, err := state.Sealed().Head()
+	if err != nil {
+		return nil, fmt.Errorf("could not retrieve last sealed block: %w", err)
+	}
+
+	core := &approvalProcessingCore{
+		approvalsCache:                       NewApprovalsCache(1000),
+		assigner:                             assigner,
+		headers:                              headers,
+		state:                                state,
+		verifier:                             verifier,
+		seals:                                seals,
+		approvalConduit:                      approvalConduit,
+		requiredApprovalsForSealConstruction: requiredApprovalsForSealConstruction,
+		emergencySealingActive:               emergencySealingActive,
+		requestTracker:                       sealing.NewRequestTracker(10, 30),
+	}
+
+	factoryMethod := func(result *flow.ExecutionResult) (*AssignmentCollector, error) {
+		return NewAssignmentCollector(result, core.state, core.headers, core.assigner, core.seals, core.verifier,
+			core.approvalConduit, core.requestTracker, core.requiredApprovalsForSealConstruction)
+	}
+
+	collectors := NewAssignmentCollectorTree(factoryMethod)
+	_, err = collectors.PruneUpToHeight(lastSealed.Height)
+	if err != nil {
+		return nil, fmt.Errorf("could not prune tree to initial height")
+	}
+
+	core.collectorTree = collectors
+
+	return core, nil
+}
+
+func (c *approvalProcessingCore) lastSealedHeight() uint64 {
+	return atomic.LoadUint64(&c.atomicLastSealedHeight)
+}
+
+func (c *approvalProcessingCore) lastFinalizedHeight() uint64 {
+	return atomic.LoadUint64(&c.atomicLastFinalizedHeight)
+}
+
+// WARNING: this function is implemented in a way that we expect blocks strictly in parent-child order
+// Caller has to ensure that it doesn't feed blocks that were already processed or in wrong order.
+func (c *approvalProcessingCore) OnFinalizedBlock(finalizedBlockID flow.Identifier) {
+	finalized, err := c.headers.ByBlockID(finalizedBlockID)
```

Comment:
⚠️ 
these two lines worry me:
https://github.com/onflow/flow-go/blob/a552e78e9b45d21c0b436eb3c6d3439b403cf6ac/engine/consensus/approvals/core.go#L109
https://github.com/onflow/flow-go/blob/a552e78e9b45d21c0b436eb3c6d3439b403cf6ac/engine/consensus/approvals/core.go#L117

Theoretically, the sealed block could increase arbitrarily in between these two queries to the protocol state. It could even become larger than the finalized height, which would leave `Core` in an internally inconsistent state. 

You can avoid this by changing the query: 
* you need `storage.Seals`
* then you can run:
   ```golang
   seal, err := seals.ByBlockID(finalizedBlockID)
   if err != nil ...
   lastSealed, err := c.headers.ByBlockID(seal.BlockID)
   ```